### DISCUSSION TITLE
Best practice implementation for DB session dependency

### DISCUSSION THREAD

**Discussion by amaschas** [fastapi/fastapi#12254] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# We are using an iterator to manage the session dependency via yield, which appears in a lot of FastAPI examples:

def db() -> Iterator[Session]:
    sess = session()
    try:
        yield sess
        sess.commit()
    except Exception:
        sess.rollback()
        raise
    finally:
        sess.close()

# And using it like this:

class Repository:
    def __init__(self, db: Session = Depends(db)):
        self._db = db

    def create(self, data) -> None:
        self._db.add(ExampleModel(**data))

    def get_by_id(self, id):
        return self._db.query(ExampleModel).filter((ExampleModel.id == id)).one_or_none()

# Problem case 1:

def do_thing(self):
    try:
      example_model = self._repo.get_by_id(1)
    except NoResultFound:
        # This exception is never caught here
        return None

# Problem case 2:

def do_thing(self):
    example_model = self._repo.get_by_id(1)
    if not example_model:
        return None

    # Instance is no longer bound to a session here
    result = example_model.value

    return example_model.id, result
```

### Description

I've been trying to find a solution to some problems that have arisen as a result of the yield pattern that FastAPI encourages users to implement for dependencies like databases. I have two specific problems I've encountered:

• Because the database connection is yielded, I can't handle that arise with database queries within the dependent function. For instance, if a database call fails, and rather than raise an error I want to handle the error and perform another operation, I can't do so.

• Session expiry is resulting in unbound objects. Again because the session is yielded to the calling function, once the call has executed the session closes and any objects returned are unbound, which means I can't access their properties.

Is there a modified version of this pattern that will address my use cases, or is this just not a use case that this pattern supports?

### Operating System

Linux

### FastAPI Version

0.109.1

### Pydantic Version

2.8.2

### Python Version

3.12.1

**Comment by ceb10n** [comment]
I would only use `Depends` to things related directly with your route rather then using in other parts of your projects.

There're two main reasons I don't like this approach:

* Your services, repositories, etc are not dependant of any web framework
* IoC/DI don't require a "*framework*" when we are developing in python

Since you didn't mention your db, I'll give you a simple example with motor/mongodb on how I would handle db connections on a FastAPI project:

`db.py`

```python
from motor.motor_asyncio import AsyncIOMotorClient


class DataBase:
    client: AsyncIOMotorClient = None


database = DataBase()


async def open_conn() -> None:
    db.client = AsyncIOMotorClient(
        your_conn_string,
        maxPoolSize={your-pool-max-size},
        minPoolSize={your-pool-min-size}
    )


async def close_conn():
    db.client.close()
```

`main.py`

```python
from contextlib import asynccontextmanager

from fastapi import FastAPI

from your_app.db import open_conn, close_conn
from your_app.services import your_service


@asynccontextmanager
async def lifespan(app: FastAPI):
    await db.open_conn()
    yield
    await db.close_conn()


app = FastAPI(lifespan=lifespan)

@app.get("/")
async def your_route():
    return await your_service.do_your_thing()

```

`your_service.py`

```python
from your_app.repositories import your_repository


async def do_your_thing() -> YourReturn:
    # do your things
    something = await your_repository.find_something()
    # do other things

    return something
```


`your_repository.py`


```python
from your_app.db import database


async def find_something():
    return database.client["your_db"]["your_collection"].find({})
```


This way your keep it simple and don't need to use any DI frameworks other than your program design.

**Comment by amaschas** [comment]
Hrm, OK. Just for context I'm using a Postgres database. So you'd suggest I not go the route documented here: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/?

**Reply by ceb10n** [reply]
> Hrm, OK. Just for context I'm using a Postgres database. So you'd suggest I not go the route documented here: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/?

Exactly.

But of course, the default answer is always: "It depends"

It depends on your scenario, your needs, the maturity of the team, etc.

From a perspective of design & architecture, I would not add a database dependency to my routes. Theoretically, your routes are responsible for handling the request, not where the data is stored.

But again, it all depends on your scenario.

And about the docs, we need to remember that the docs are about the FastAPI features and how you can use them.

If Sebastián were to write all the development standards and best practices in the documentation, he wouldn't have time for anything else. And of course, that's not the point when writing documentation for a framework.

**Comment by krylosov-aa** [comment]
Hi! Maybe my approach to working with sessions will be useful. It's done through context without dependencies. I'm currently using it with my FastApi app running Postgres.

https://github.com/krylosov-aa/context-async-sqlalchemy
